### DISCUSSION TITLE
MapStruct Ambiguous constructors

### DISCUSSION THREAD

**Discussion by RodrigoMonterio** [mapstruct/mapstruct#3590] [General]
I have a problem using MapStruct. I created a record that has an id of type long and in my mapper I need to convert this object to the Long type, but the error below occurs. I will leave the code to make the problem clearer, sorry if any information is missing, it's my first question on the forum.

Ambiguous constructors found for creating java.lang.Long: Long(long), Long(java.lang.String). Either declare parameterless constructor or annotate the default constructor with an annotation named @Default.

My code:

```java
//IdClient.java

public record IdClient(Long id) { }
```
```java
//ClientEntity.java

@Table(name = "client") @Builder 
public record ClientEntity(Long idClient){} 
```

```java
//ClientMapper.java

@ApplicationScoped public interface ClientMapper { 

  ClientEntity IdClientToClientEntity(IdClient id);

  Long map(IdClient id);

}
```

I solved it, `@Mapping(target = "idClient", source = "idClient.id")` but I'm not sure if this is the most appropriate way, considering that the record is of type long and my entity expects a long. I opened this topic to find out what would be the best approach for this case.

**Comment by zyberzebra** [comment]
`@Mapping(target = "idClient", source = "idClient.id")` is the correct solution.
if your records would be:
```java
public record IdClient(Long id) { }
public record ClientEntity(Long id){} 
``` 
The Mapper would map automatically without you specifying the Mapping via `@Mapping`

See: [Basic Mapping](https://mapstruct.org/documentation/stable/reference/html/#basic-mappings)

> When a property has the same name as its target entity counterpart, it will be mapped implicitly.
When a property has a different name in the target entity, its name can be specified via the @Mapping annotation.

**Reply by RodrigoMonterio** [reply]
Understood, thank you for your explanation @zyberzebra
